### PR TITLE
opds: add opensearch support

### PR DESCRIPTION
**Pull request by philips**
Tested with koreader and works great!

Discussed with original OPDS author @Scotsguy here

https://github.com/9001/copyparty/pull/779#issuecomment-3482160504

Bassed on these specs

https://specs.opds.io/opds-1.2#3-search
https://github.com/koreader/koreader/pull/7380

To show that your contribution is compatible with the MIT License, please include the following text somewhere in this PR description:  
This PR complies with the DCO; https://developercertificate.org/  

Demo video of Koreader

### CODE REVIEW

**Comment by 9001** [comment]
nice! spotted a tiny thing that I'll fix real quick in a sec, other than that good stuff, thx o/

**Comment by philips** [comment]
> nice! spotted a tiny thing that I'll fix real quick in a sec, other than that good stuff, thx o/

Woah, thanks for the fast review. Happy to fix issues. This is my first time in the code base and not 100% confident.

**Comment by Scotsguy** [comment]
Nice! 
- Did you test this with readers other than koreader? I tested with moon reader (android) when i implemented opds initially, spotted a few corner cases that way
- How does this interact with dirkeys? It looks like you're synthesizing URLs that may not have the necessary dirkeys in them
- The hardcoded "Copyparty" string should ~~arguably be the hostname (unless the config option to hide it is set)~~ (edit: this already includes bname, didn't see it) or at least respect `-nb`

**Comment by 9001** [comment]
Alright, made some small tweaks for everything I could notice -- some tiny optimizations and reformatted with black, and url-encoded the prefix so it'll also work in paths with special characters. So just nitpicks really, once again good stuff!

I think the URLs *should* be fine; it mostly uses the ones that are produced by `idx.search` which at least has the filekeys, unless I missed something.

if you guys think this looks good then I'm ready to merge whenever :+1:

File: copyparty/web/opds.xml [hunk 1]
```diff
@@ -1,5 +1,9 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <feed xmlns="http://www.w3.org/2005/Atom">
+    <link rel="search"
+    href="{{ opds_osd | e }}"
+    type="application/opensearchdescription+xml"
+    title="Search"/>
```
Comment (Scotsguy) [inline copyparty/web/opds.xml]: For the sake of being language-agnostic, is this title necessary? Will clients synthesize a name when the title isn't specified, or will they ignore the search description instead?

The OPDS spec is (of course) light on details, but they don't seem to include the title in their example link

**Comment by philips** [comment]
@Scotsguy 

> * Did you test this with readers other than koreader? I tested with moon reader (android) when i implemented opds initially, spotted a few corner cases that way

Tested with Moon Reader and caught an escaping bug fixed in 70ad121

> * How does this interact with dirkeys? It looks like you're synthesizing URLs that may not have the necessary dirkeys in them

I am not familiar with dirkeys but @9001 said it is OK? 

https://github.com/9001/copyparty/pull/1287#issuecomment-3874150680

> * The hardcoded "Copyparty" string should ~arguably be the hostname (unless the config option to hide it is set)~ (edit: this already includes bname, didn't see it) or at least respect `-nb`

Removed in 0342506

File: copyparty/web/opds.xml (hunk 1 above)
Comment (philips) [reply copyparty/web/opds.xml]: Removed in 1c8d37d. It seems Koreader and Moon Reader don't care.

**Comment by philips** [comment]
I think this is good to merge if @Scotsguy likes the patches to address feedback. It is working for me with Koreader and Moon Reader on my server.

File: copyparty/web/opds_osd.xml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<OpenSearchDescription xmlns="http://a9.com/-/spec/opensearch/1.1/">
+  <ShortName>{{ longname | truncate(14) | e }}</ShortName>
```
Comment (Scotsguy) [inline copyparty/web/opds_osd.xml]: this can be truncate(16) now

**Comment by Scotsguy** [comment]
I just have the one nitpick, otherwise LGTM

**Comment by philips** [comment]
@Scotsguy Fixed in [01cbf45](https://github.com/9001/copyparty/commit/01cbf4592076eaff57f325e645295df72dd6f40a)

File: copyparty/web/opds_osd.xml (hunk 2 above)
Comment (philips) [reply copyparty/web/opds_osd.xml]: @Scotsguy Fixed in [01cbf45](https://github.com/9001/copyparty/commit/01cbf4592076eaff57f325e645295df72dd6f40a)

**Comment by philips** [comment]
@9001 I think we are ready to merge. Got the 'ol thumbs up on my fix from @Scotsguy in https://github.com/9001/copyparty/pull/1287#issuecomment-3885020706

**Comment by 9001** [comment]
sweet! Thanks again o/
